src/systems/filecoin_markets/storage_market/storage_market_actor_state.go

```diff
@@ -0,0 +1,288 @@
+func (st *StorageMarketActorState_I) _transferBalance(rt Runtime, fromLocked addr.Address, toAvailable addr.Address, amount actor.TokenAmount) {
+	fromB := st.Balances()[fromLocked]
+	toB := st.Balances()[toAvailable]
+
+	if fromB.Locked() < amount {
+		rt.Abort("sma._transferBalance: attempt to unlock funds greater than actor has")
+		return
+	}
+
+	fromB.Impl().Locked_ -= amount
+	toB.Impl().Available_ += amount
+}
+
+func (st *StorageMarketActorState_I) _lockFundsForStorageDeal(rt Runtime, deal deal.StorageDeal) {
+	p := deal.Proposal()
+
+	st._lockBalance(rt, p.Client(), p.ClientBalanceRequirement())
+	st._lockBalance(rt, p.Provider(), p.ProviderBalanceRequirement())
+}
+
+func (st *StorageMarketActorState_I) _getOnchainDeal(rt Runtime, dealID deal.DealID) deal.OnchainDeal {
+	deal, found := st.Deals()[dealID]
+	if !found {
+		rt.Abort("sm._getOnchainDeal: dealID not found in Deals.")
+	}
+
+	return deal
+}
+
+func (st *StorageMarketActorState_I) _assertPublishedDealState(rt Runtime, dealID deal.DealID) {
+
+	deal := st._getOnchainDeal(rt, dealID)
+
+	if deal.LastPaymentEpoch() != block.ChainEpoch(0) {
+		rt.Abort("sma._assertPublishedDealState: deal is not in PublishedDealState.")
+	}
+
+}
+
+func (st *StorageMarketActorState_I) _assertActiveDealState(rt Runtime, dealID deal.DealID) {
+
+	deal := st._getOnchainDeal(rt, dealID)
+
+	if deal.LastPaymentEpoch() <= block.ChainEpoch(0) {
+		rt.Abort("sma._assertActiveDealState: deal is not in ActiveDealState.")
+	}
+}
+
+func (st *StorageMarketActorState_I) _getParticipantBalance(rt Runtime, participant addr.Address) StorageParticipantBalance {
+	balance, found := st.Balances()[participant]
+
+	if !found {
+		rt.Abort("sma._getParticipantBalance: participant balance not found.")
+	}
+
+	return balance
+}
+
+func (st *StorageMarketActorState_I) _getStorageFeeSinceLastPayment(rt Runtime, deal deal.OnchainDeal, newPaymentEpoch block.ChainEpoch) actor.TokenAmount {
+
+	duration := newPaymentEpoch - deal.LastPaymentEpoch()
+	dealP := deal.Deal().Proposal()
+	fee := actor.TokenAmount(0)
+
+	if duration > 0 {
+		unitPrice := dealP.StoragePricePerEpoch()
+		fee := actor.TokenAmount(uint64(duration) * uint64(unitPrice))
+
+		clientBalance := st._getParticipantBalance(rt, dealP.Client())
+
+		if fee > clientBalance.Locked() {
+			rt.Abort("sma._getStorageFeeSinceLastPayment: fee cannot exceed client LockedBalance.")
+		}
+
+	} else {
+		rt.Abort("sma._getStorageFeeSinceLastPayment: no new payment since last payment.")
+	}
+
+	return fee
+
+}
+
+func (st *StorageMarketActorState_I) _slashDealCollateral(rt Runtime, dealP deal.StorageDealProposal) {
+	amountToSlash := dealP.ProviderBalanceRequirement()
+
+	providerBal, found := st.Balances()[dealP.Provider()]
+	if !found {
+		rt.Abort("sma._slashDealCollateral: provider not found in balances.")
+	}
+
+	if providerBal.Locked() < amountToSlash {
+		amountToSlash = providerBal.Locked()
+		// TODO: decide on error handling here
```

Comment:
slash it all anyways and log error?

Reply:
yes, that is the intention. Will fix in future PR with better error handling.